File: 2022/10.cc
#include "common.h"

namespace aoc_2022_10 {

struct Instruction {
    int due;
    int addend;
};

void run(std::string_view buf)
{
    int X = 1;
    int cycle = 1;
    int signal_strength = 0;
    std::string crt;
    Instruction pending{-1, 0};

    auto lines = split_lines(buf);
    for (size_t i = 0; i < lines.size();) {
        // Start of the cycle. If nothing is executing, fetch an instruction.
        if (pending.due < 0) {
            auto &line = lines[i++];
            if (line == "noop") {
                pending = {cycle, 0};
            } else {
                int addend = 0;
                std::from_chars(line.data() + 5, line.data() + line.size(), addend);
                pending = {cycle + 1, addend};
            }
        }

        // Get signal strength for part 1:
        if (cycle % 40 == 20)
            signal_strength += cycle * X;

        // Draw to CRT for part 2:
        const int col = (cycle - 1) % 40;
        if (!crt.empty() && col == 0)
            crt += '\n';
        crt += " #"[X - 1 <= col && col <= X + 1];

        // End of the cycle.
        if (pending.due == cycle) {
            // Current instruction is due; finish it.
            X += pending.addend;
            pending.due = -1;
        }

        cycle++;
    }

    fmt::print("{}\n", signal_strength);
    fmt::print("{}\n", crt);
}

}

File: 2022/10.py
#!/usr/bin/env python3
import sys


INSTR_TIMINGS = {
    "noop": 0,
    "addx": 1,
}


def main():
    X = 1
    cycle = 1
    pending_instr = None
    signal_strength = 0
    crt_rows = []

    # Main loop; 1 iteration per cycle
    lines = sys.stdin.readlines()
    while lines:
        # Start of the cycle. If nothing is executing, fetch an instruction.
        if pending_instr is None:
            line = lines.pop(0)
            op, *args = line.strip().split()
            pending_instr = (cycle + INSTR_TIMINGS[op], op, *map(int, args))

        # Get signal strength for part 1:
        if cycle in (20, 60, 100, 140, 180, 220):
            signal_strength += cycle * X

        # Draw to CRT for part 2:
        col = (cycle - 1) % 40
        if col == 0:
            # Push a new scanline
            crt_rows.append("")
        crt_rows[-1] += " #"[X - 1 <= col <= X + 1]

        # End of the cycle.
        if pending_instr[0] == cycle:
            # Current instruction is due; finish it.
            match pending_instr:
                case (_, "addx", arg):
                    X += arg
            pending_instr = None

        cycle += 1

    print(signal_strength)
    print("\n".join(crt_rows))


if __name__ == "__main__":
    main()
